### ISSUE TITLE
Escaping of single `!` on Postgres, MySQL and SQLite

### ISSUE THREAD

**Issue by do4gr** [jOOQ/jOOQ#8422]
### Expected behavior and actual behavior:

We have a table Comment with a column text. The table has one row with the entry `IMPORTANT!` for text. 

We want to query for all Comments that end with `!`.

For Postgres jOOQ generates: 

```
select "text" 
	from "Comment" as C	
	where C."text" like concat(
    '%', 
    '!'
  ) escape '!';
```
This errors with `LIKE pattern must not end with escape character`. Leaving out `escape '!'` would return the correct result. 

For MySQL jOOQ generates: 

```
select C.`text` 
	from `Comment` as C	
	where C.`text` like concat('%' || '!') escape '!';
```
This does not error but returns 0 rows. Using `C."text" like '%!'` returns the correct result for MySQL.

For SQLite jOOQ generates: 

```
select C.`text` 
	from `Comment` as C	
	where C.`text`like ('%' || '') escape '!';
```
This does not error but returns 0 rows as well.

- jOOQ: 3.11.9
- Scala: 2.12.7
- Database (include vendor): MySQL, PostGres, SQLite
- OS: MacOs
- JDBC Driver (include name if inofficial driver):
  mariaDbClient  = "org.mariadb.jdbc"  % "mariadb-java-client" % "2.1.2"
  postgresClient = "org.postgresql"    % "postgresql"          % "42.2.2"
  sqliteClient   = "org.xerial"        % "sqlite-jdbc"         % "3.25.2"

**Comment by lukaseder** [comment]
Thanks for your report. Would you mind adding the jOOQ code that produces these queries?

**Comment by do4gr** [comment]
We're composing our queries so they're not in one place. I guess the interesting part for you would be the condition part. 

There we're generating a Field[AnyRef] and then appending different filters to build a condition, in this case endsWith. One peculiarity of our setup is that we provide dummyValues while constructing the query and then overwrite the positioned parameters later. So in this case we provide an empty String to jOOQ and then later set the PP to `!`.

So to jOOQ the condition would be field[AnyRef].endsWith("")

I hope this helps. 

This is where the conditions are build: https://github.com/prisma/prisma/blob/d7101eef14133b53c4463994192205d54f326105/server/connectors/api-connector-jdbc/src/main/scala/com/prisma/api/connector/jdbc/database/FilterConditionBuilder.scala#L60

**Comment by lukaseder** [comment]
Thanks for the explanation. I see, this is because we're eagerly escaping the argument you're passing to `endsWith()` without taking into account that the actual bind variable could be replaced when the query is executed, in case of which no further escaping takes place.

We could escape the actual value just when it is being bound to the JDBC statement (difficult to get right in the presence of custom data type bindings), or just revert to escaping the value in SQL.

I've created a separate issue to fix this: #8429. Closing this issue as a duplicate of #8429.

Thanks again for your report.

**Comment by lukaseder** [comment]
The workaround for you is to not use `endsWith()` and similar convenience methods, and implement the *"ends with"* semantics manually using `LIKE`

**Comment by knutwannheden** [comment]
@do4gr: Would you be able to provide an MCVE (https://github.com/jOOQ/jOOQ-mcve) to help reproduce this? We would like to make sure that we have your exact use case covered.

**Comment by knutwannheden** [comment]
Meanwhile the separate issue #8429 has been closed and will be delivered as part of the 3.12 release. We expect this should resolve your problem.

**Comment by do4gr** [comment]
We'll upgrade to 3.12 as soon as it's out and report back then.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.12 (#8429). Backport scheduled for 3.11.12 (#8525)
